openshift/machine-config-operator
# Add kubens.service, drop-ins, and kubensenter prefix to kubelet.service

**- What I did**

In support of openshift enhancement https://github.com/openshift/enhancements/blob/master/enhancements/hide-container-mountpoints.md

This comes in 2 parts:

- One prefixes the kubelet.service files so they start via the new 'kubensenter' script.
- The other adds the new 'kubens.service' and related drop-ins for kubelet and cri-o.

The new 'kubens.service' that drives this new feature is disabled by default, so should be safe to go in as soon as the openshift-hyperkube RPM lands which includes the new 'kubensenter' script required for the 1st part. (https://github.com/openshift/kubernetes/pull/1327)

/hold until openshift-hyperkube arrives with the new kubensenter script.

**- How to verify it**

This cannot be mergedf until the new `kubensenter` script arrives via an updated openshift-hyperkube, but enabling of this feature also requires that a CRI-O change (https://github.com/cri-o/cri-o/pull/5974) that will arrive once RHCOS includes CRI-O 1.25.

Once both prerequisites are in place, this feature should be verified as follows:
- Initially this must have NO EFFECT unless enabled.  Systemd, kubelet, and cri-o are all in the same mount namespace as usual:

```
[root@node ~]# readlink /proc/1/ns/mnt
mnt:[4026531840]
[root@node ~]# readlink /proc/$(pgrep kubelet)/ns/mnt
mnt:[4026531840]
[root@node ~]# readlink /proc/$(pgrep crio)/ns/mnt
mnt:[4026531840]
```
Note all 3 are identical

- Enabling the feature can be done by injecting a single MachineConfig object that enables the new `kubens.service` systemd service.  If this is done, both kubelet and CRI-O should be in a different namespace that systemd, but the system should otherwise behave normally:

With the feature enabled:
```
[root@node ~]# readlink /proc/1/ns/mnt
mnt:[4026531840]
[root@node ~]# readlink /proc/$(pgrep kubelet)/ns/mnt
mnt:[4026534696]
[root@node ~]# readlink /proc/$(pgrep crio)/ns/mnt
mnt:[4026534696]
```
Note that kubelet and crio are identical, but unique from pid 1

**- Description for the changelog**

Adds a new optional systemd service called `kubens.service` which can be enabled to segregate all Kubernetes-specific mount points into a new mount namespace separate from the host OS and systemd.
